### PR TITLE
Adds project quickjs

### PR DESCRIPTION
**Pull request by catenacyber**
Follows #3473 with approval from project's maintainer Fabirce Bellard by mail (and fix in API use) :

> Pas de problème bien sûr pour l'intégration de QuickJS à oss-fuzz

cc @Dor1s

### CODE REVIEW

File: projects/quickjs/build.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# Makefile should not override CFLAGS
+sed -i -e 's/CFLAGS=/CFLAGS+=/' Makefile
+CONFIG_CLANG=y make libquickjs.a
+zip -r $OUT/fuzz_eval_seed_corpus.zip tests/*.js
+zip -r $OUT/fuzz_eval_seed_corpus.zip examples/*.js
+zip -r $OUT/fuzz_compile_seed_corpus.zip tests/*.js
+zip -r $OUT/fuzz_compile_seed_corpus.zip examples/*.js
+
+cd ..
+$CC $CFLAGS -Iquickjs -c fuzz_eval.c -o fuzz_eval.o
```
Comment (Dor1s) [inline projects/quickjs/build.sh]: if there is any reasonable seed corpus for the regexp fuzzer, you can write a small loop here, e.g.

```sh
FUZZ_TARGETS="fuzz_eval fuzz_compile fuzz_regexp"
for f in $FUZZ_TARGETS; do
  # $CC etc to compile $f
  # copy seed corpus to $OUT
done
```

that would be handy when you add more fuzz targets and have to change the compilation command

File: projects/zlib-ng/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 Google Inc.
+# Copyright 2020 Google Inc.
```
Comment (Dor1s) [inline projects/zlib-ng/Dockerfile]: please revert changes in this file, seems accidental

File: projects/quickjs/project.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+homepage: "https://bellard.org/quickjs/"
+primary_contact: "fabrice@bellard.org"
+auto_ccs :
+- "p.antoine@catenacyber.fr"
+
+sanitizers:
+- address
```
Comment (Dor1s) [inline projects/quickjs/project.yaml]: any particular reason not to include ubsan? if not, just remove `sanitizers` section and both ASan and UBSan will be used by default

File: projects/quickjs/build.sh (hunk 1 above)
Comment (catenacyber) [reply projects/quickjs/build.sh]: > there is any reasonable seed corpus for the regexp fuzzer

Not yet

But I added the loop as suggested, thanks

File: projects/zlib-ng/Dockerfile (hunk 2 above)
Comment (catenacyber) [reply projects/zlib-ng/Dockerfile]: Oops, should be fixed

File: projects/quickjs/project.yaml (hunk 3 above)
Comment (catenacyber) [reply projects/quickjs/project.yaml]: I remembered there was a problem with it, but not which one, let's try again with CI

**Comment by catenacyber** [comment]
So, problem with UBSAN is use of `uint8_t str8[0]` in structure `JSString` producing error
` runtime error: index 4 out of bounds for type 'uint8_t [0]'`

It is not a problem for ASAN as structure `JSString` is allocated with more space to contain the 5 characters.

This is a trick to avoid doing 2 allocations instead of only one.

Is there some `UBSAN_OPTIONS` value to allow this ?

**Comment by Dor1s** [comment]
> Is there some `UBSAN_OPTIONS` value to allow this ?

I think compiling with `-fno-sanitize=array` should hide these

**Pull request merged**